File: MoonEngine-Root/OpenGL/src/main/java/com/greentree/engine/moon/opengl/adapter/FreamBufferRenderTarget.java
package com.greentree.engine.moon.opengl.adapter;

import com.greentree.common.graphics.sgl.freambuffer.FreamBuffer;
import com.greentree.commons.graphics.smart.texture.Texture;
import com.greentree.engine.moon.render.pipeline.target.RenderTarget;
import com.greentree.engine.moon.render.pipeline.target.RenderTargetTextute;
import com.greentree.engine.moon.render.pipeline.target.buffer.TargetCommandBuffer;

import java.util.Objects;

public record FreamBufferRenderTarget(FreamBuffer framebuffer, RenderTarget context)
        implements RenderTargetTextute {

    public FreamBufferRenderTarget {
        Objects.requireNonNull(framebuffer);
        Objects.requireNonNull(context);
    }

    @Override
    public TargetCommandBuffer buffer() {
        return new FreamBufferCommandBuffer(framebuffer, context.buffer());
    }

    @Override
    public Texture getColorTexture(int index) {
        final var texture = framebuffer.getColorTexture(index);
        return new OpenGLTexture(texture);
    }

    @Override
    public Texture getDepthTexture() {
        final var texture = framebuffer.getDepthTexture();
        return new OpenGLTexture(texture);
    }

}

File: MoonEngine/OpenGL/src/main/java/com/greentree/engine/moon/opengl/adapter/FreamBufferRenderTarget.kt
package com.greentree.engine.moon.opengl.adapter

import com.greentree.common.graphics.sgl.freambuffer.FreamBuffer
import com.greentree.commons.graphics.smart.target.FrameBuffer
import com.greentree.commons.graphics.smart.target.RenderCommandBuffer
import com.greentree.commons.graphics.smart.target.RenderTarget
import com.greentree.commons.graphics.smart.texture.Texture
import java.util.*

data class FreamBufferRenderTarget(val framebuffer: FreamBuffer, val context: RenderTarget) : FrameBuffer {

	override fun buffer(): RenderCommandBuffer {
		return FreamBufferCommandBuffer(framebuffer, context.buffer())
	}

	override fun getColorTexture(index: Int): Texture {
		val texture = framebuffer.getColorTexture(index)
		return OpenGLTexture(texture)
	}

	override fun getDepthTexture(): Texture {
		val texture = framebuffer.depthTexture
		return OpenGLTexture(texture)
	}

	init {
		Objects.requireNonNull(framebuffer)
		Objects.requireNonNull(context)
	}
}